py/foundation/document.py

```diff
@@ -1,33 +1,67 @@
 """ Foundation document structures. """
 
-from dataclasses import dataclass, field
+import json
+
+from dataclasses import asdict, dataclass, field, replace
+from functools import lru_cache
 from itertools import chain
-from typing import Tuple, Iterable, Type, TypeVar
+from pathlib import Path
+from typing import Dict, Optional, Iterable, Tuple, Type, TypeVar
 
 from .entity import Entity
 from .geometry import BBox
+from .ocr import InputWord
 from .typing_utils import unwrap
+from ._instantiate import _instantiate
+
 
 E = TypeVar('E', bound=Entity)
 
 
 @dataclass(frozen=True)
 class Document:
-  """A Foundation Document is a collection of Entity.
-  """
+  """A Foundation Document is a collection of Entities of varying type."""
   bbox: BBox
   entities: Tuple[Entity, ...] = field(default_factory=tuple)
+  name: Optional[str] = None
```

Comment:
What is name here? Is that like the name of a potential file?

Reply:
Yeah. In blueprint before, we would say things like `document.input_doc.name` to get the file name, and now that we won't have the customized BP `input_doc` field on Document anymore, this is to be able to keep track of that if wanted.